### ISSUE TITLE
"Realm accessed from incorrect thread." when using semaphore

### ISSUE THREAD

**Issue by tomdaniel0x01** [realm/realm-cocoa#5880]
## Goals, Expected Results, Actual Results, Steps to Reproduce
I have a Realm List that can become several GiB in size. I am trying split a Realm List into segments and then process each segment one at a time. I need to take out 8MiB of data from the list, convert to JSON, then upload to a Firebase function, and repeat this until the Realm List has been completely processed. I cannot load the entire list into memory for processing because several GiB is too much to load into memory at once. This means I should only convert 8MiB at a time into JSON, then wait until processing is complete and the JSON has been deallocated from memory before continuing processing. I was trying to use semaphores for this, but it results in the error "Realm accessed from incorrect thread". If I comment out "semaphore.wait()" the error does not occur.

It crashed at the line:
> var jsonData = try? JSONEncoder().encode(segment)

## Code Sample
```
        let uploadCache = List<AnyDataSlice>()
        let semaphore = DispatchSemaphore(value: 1)

        uploadCache.append(objectsIn: dataSlices)
        while !uploadCache.isEmpty {
            finishCount += 1
            DispatchQueue.global(qos: .background).async {
                semaphore.wait()
                let segment = List<AnyDataSlice>()
                if uploadCache.count > segmentSize {
                    segment.append(objectsIn: uploadCache.dropLast(uploadCache.count - segmentSize))
                } else {
                    segment.append(objectsIn: uploadCache)
                }
                let range = segment.startIndex..<segment.endIndex
                uploadCache.removeSubrange(range)
                var jsonData = try? JSONEncoder().encode(segment)
                if jsonData == nil { return }
                var jsonString = String(data: jsonData!, encoding: .utf8)
                if jsonString == nil { return }
                jsonData = nil

                // Upload JSON string to Firebase function

                jsonString = nil
                self.semaphore.signal()
            }
        }
```

## Version of Realm and Tooling
```
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G65

/Applications/Xcode.app/Contents/Developer
Xcode 9.4.1
Build version 9F2000

/usr/local/bin/pod
1.5.3
Realm (3.7.5)
RealmSwift (3.7.5)

/bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin17)


(not in use here)

/usr/bin/git
git version 2.15.2 (Apple Git-101.1)
```

**Comment by tomdaniel0x01** [comment]
Here is my currently working workaround. It's a little messier than semaphors.


```
class DataRepository {
    // MARK: Singleton shared instance
    //
    static let shared = DataRepository()
    
    var realm = try! Realm()
    
    // MARK: background
    //
    func background(sync: Bool = false, action: @escaping (_: Realm) -> ()) {
        if sync {
            DispatchQueue.global(qos: .background).sync {
                autoreleasepool {
                    let realm = try! Realm()
                    action(realm)
                }
            }
        } else {
            DispatchQueue.global(qos: .background).async {
                autoreleasepool {
                    let realm = try! Realm()
                    action(realm)
                }
            }
        }
    }
}
```
```
self!.dataRepository.background(sync: true) { (realm) in
                if let persistedCache = self!.dataRepository.get(DataSliceCache.self, forPrimaryKey: "dataSliceCache", instance: realm),
                    persistedCache.dataSlices.count != 0
                {
                    debugPrint("Persisted data slices: \(persistedCache.dataSlices.count)")
                    var hasRun = false
                    var segmentedCache = self!.segment(dataSlices: persistedCache.dataSlices, segmentSize: maximumDataSliceArraySize)
                        
                    func uploadSegment() {
                        let nextSegment = segmentedCache.removeFirst()
                        self!.upload(dataSlices: nextSegment) { (result, error) in
                            if let error = error {
                                debugPrint(error)
                                return
                            }
                            
                            if let status = (result?.data as? [String: String])?["status"] {
                                debugPrint("[Firebase function - logData] Status: \(status)")
                                if status == "success" {
                                    // Delete this segment from persisted cache
                                    //
                                    do {
                                        try self!.dataRepository.delete(nextSegment, instance: realm)
                                        debugPrint("Successfully removed data slice segment from persisted storage")
                                    } catch {
                                        debugPrint("Failed to delete data slices with error - \(error)")
                                    }
                                }
                            }
                            
                            if !segmentedCache.isEmpty {
                                uploadSegment()
                            }
                        }
                    }
                    
                    if !hasRun {
                        hasRun = true
                        uploadSegment()
                    }
                }
            }
```
```
func segment(dataSlices: List<AnyDataSlice>, segmentSize: Int) -> Array<List<AnyDataSlice>>{
        let dataSliceCache = List<AnyDataSlice>()
        var segmentedCache = [List<AnyDataSlice>]()
        dataSliceCache.append(objectsIn: dataSlices)

        while !dataSliceCache.isEmpty {
            let segment = List<AnyDataSlice>()
            if dataSliceCache.count > segmentSize {
                segment.append(objectsIn: dataSliceCache.dropLast(dataSliceCache.count - segmentSize))
            } else {
                segment.append(objectsIn: dataSliceCache)
            }
            segmentedCache.append(segment)
            let range = segment.startIndex..<segment.endIndex
            dataSliceCache.removeSubrange(range)
        }
        
        return segmentedCache
    }
```
```
func upload(dataSlices: List<AnyDataSlice>, completion: @escaping (HTTPSCallableResult?, Error?) -> ()) {
        let numberOfDataSlices = dataSlices.count
        guard let jsonData = try? jsonEncoder.encode(dataSlices) else { return }
        let jsonString = String(data: jsonData, encoding: .utf8)
        debugPrint("Data slice payload is \(jsonString!.lengthOfBytes(using: .utf8)) bytes containing \(numberOfDataSlices) data slice(s)")
        debugPrint("Uploading data slice payload...")
        functions.httpsCallable("logData").call(jsonString!, completion: completion)
    }
```

**Comment by tomdaniel0x01** [comment]
I found another error caused by semaphores which is probably related to this one. I am executing a Realm write and add on a background thread, then creating a thread safe reference to the newly managed object. I then have a callback that executes on the originating thread before the Realm write command was executed on a background thread. I am using a semaphore to pause the execution of the callback until the write is complete, then signaling the semaphore which allows the callback to execute. In the callback I am attempting to resolve the thread safe reference to the object that was added and received the following error:

`Terminating app due to uncaught exception 'RLMException', reason: 'std::exception'`
